### PR TITLE
drivers: usb: udc_dwc2: Support Remote Wakeup

### PR DESCRIPTION
**Pull request by tmon-nordic**
Rework bottom-half processing to improve hibernation handling necessary for proper remote wakeup from hibernation handling.

### CODE REVIEW

File: drivers/usb/udc/udc_dwc2.c [hunk 1]
```diff
@@ -842,9 +847,58 @@ static uint8_t pull_next_ep_from_bitmap(uint32_t *bitmap)
 	}
 }
 
+static void dwc2_handle_hibernation_exit(const struct device *dev,
+					 bool remote_wakeup, bool bus_reset)
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_dwc2.c:851]: Please use rwup abbreviation.
```
static void dwc2_handle_hibernation_exit(const struct device *dev,
					 const bool rwup, const bool bus_reset)
```

File: drivers/usb/udc/udc_dwc2.c [hunk 2]
```diff
@@ -130,9 +133,11 @@ struct udc_dwc2_data {
 };
 
 static void dwc2_on_bus_reset(const struct device *dev);
-static void dwc2_restore_device_registers(const struct device *dev);
+static void dwc2_restore_device_registers(const struct device *dev,
+					  bool remote_wakeup);
 static void dwc2_enter_hibernation(const struct device *dev);
-static void dwc2_exit_hibernation(const struct device *dev);
+static void dwc2_exit_hibernation(const struct device *dev, bool remote_wakeup);
+static void cancel_hibernation_request(struct udc_dwc2_data *const priv);
```
Comment (jfischer-no) [inline drivers/usb/udc/udc_dwc2.c:140]: Please get rid of terrible forward definitions.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/udc/Kconfig.dwc2 [hunk 3]
```diff
@@ -37,11 +37,3 @@ config UDC_DWC2_THREAD_PRIORITY
 	default 8
```
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply drivers/usb/udc/Kconfig.dwc2:37]: One k_event is able to handle up to 32 events. For a fully maxed out number of endpoints, there are 32 bits needed - 1 set for transfers queued and 1 set for transfers completed. The third k_event is for the overall.

However, take a look at the actual resource usage. The three k_events do consume less memory and do provide faster handling than the single k_queue. There are absolutely no benefits IMHO for the k_queue and a very serious drawback (losing messages on queue full, which is more likely at High-Speed due to 40x the bandwidth compared to Full-Speed).

File: drivers/usb/udc/udc_dwc2.c [hunk 4]
```diff
@@ -924,7 +950,19 @@ static ALWAYS_INLINE void dwc2_thread_handler(void *const arg)
...
 				dwc2_on_bus_reset(dev);
+			} else {
+				/* Resume any pending transfer handling */
+				if (k_event_test(&priv->xfer_evt, UINT32_MAX)) {
+					k_event_post(&priv->drv_evt, BIT(DWC2_DRV_EVT_XFER));
+				}
+
+				if (k_event_test(&priv->ep_finished_evt, UINT32_MAX)) {
+					k_event_post(&priv->drv_evt, BIT(DWC2_DRV_EVT_EP_FINISHED));
+				}
```
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply drivers/usb/udc/udc_dwc2.c]: It is in `dwc2_handle_hibernation_exit()` moved in a later commit in the series (your question already appears as "Outdated")

File: drivers/usb/udc/udc_dwc2.c (hunk 2 above)
Comment (tmon-nordic) [reply drivers/usb/udc/udc_dwc2.c]: Added new commit reordering the functions prior to all the changes here.

File: drivers/usb/udc/udc_dwc2.c (hunk 1 above)
Comment (tmon-nordic) [reply drivers/usb/udc/udc_dwc2.c]: Changed, but I don't quite like `rwup`. Saving few characters is not worth all the "what this rwup even means?" that will inevitably happen.

File: drivers/usb/udc/udc_dwc2.c [hunk 5]
```diff
@@ -128,6 +130,7 @@ struct udc_dwc2_data {
 };
 
 static void dwc2_on_bus_reset(const struct device *dev);
+static void dwc2_enter_hibernation(const struct device *dev);
```
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply drivers/usb/udc/udc_dwc2.c]: Reordered the handlers to get rid of all forward declarations in the file.

File: drivers/usb/udc/udc_dwc2.c [hunk 6]
```diff
@@ -111,6 +93,12 @@ struct dwc2_reg_backup {
 /* Driver private data per instance */
 struct udc_dwc2_data {
 	struct k_thread thread_data;
+	/* Main events the driver thread waits for */
+	struct k_event drv_evt;
+	/* Transfer triggers (OUT on bits 0-15, IN on bits 16-31) */
+	struct k_event xfer_evt;
+	/* Finished transactions (OUT on bits 0-15, IN on bits 16-31) */
+	struct k_event ep_finished_evt;
```
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply drivers/usb/udc/udc_dwc2.c]: Renamed.

File: drivers/usb/udc/Kconfig.dwc2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfischer-no) [reply drivers/usb/udc/Kconfig.dwc2:37]: There should be `select EVENTS` between L:9 and L:10

File: drivers/usb/udc/udc_dwc2.c [hunk 7]
```diff
@@ -2695,60 +2683,111 @@ static void udc_dwc2_isr_handler(const struct device *dev)
...
+
+			if (!udc_ep_is_busy(dev, ep_cfg->addr)) {
+				dwc2_handle_xfer_next(dev, ep_cfg);
+			} else {
+				LOG_DBG("ep 0x%02x busy", ep_cfg->addr);
+			}
+		}
+	}
+
+	if (evt & BIT(DWC2_DRV_EVT_SETUP)) {
```
[the remark this answers is not in the document]
Comment (tmon-nordic) [reply drivers/usb/udc/udc_dwc2.c]: Changed.

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**